### PR TITLE
Fix/sentinel filter

### PR DESCRIPTION
**Pull request by r27153733**
#2620

### CODE REVIEW

File: filter/sentinel/filter.go [hunk 1]
```diff
@@ -110,6 +101,12 @@ type sentinelProviderFilter struct{}
 
 func newSentinelProviderFilter() filter.Filter {
 	if sentinelProvider == nil {
+		initOnce.Do(func() {
+			err := sentinel.InitDefault()
+			if err != nil {
+				panic(err)
+			}
```
Comment (justxuewei) [inline filter/sentinel/filter.go:108]:
```suggestion
if err := sentinel.InitDefault(); err != nil {
    panic(err)
}
```

File: filter/sentinel/filter.go [hunk 2]
```diff
@@ -129,8 +126,9 @@ func (d *sentinelProviderFilter) Invoke(ctx context.Context, invoker protocol.In
 	if b != nil {
 		// interface blocked
 		return sentinelDubboProviderFallback(ctx, invoker, invocation, b)
+	} else {
+		defer interfaceEntry.Exit()
 	}
```
Comment (justxuewei) [inline filter/sentinel/filter.go:131]: `else` here could be removed.

```suggestion
	}
	defer interfaceEntry.Exit()
```

File: filter/sentinel/filter.go [hunk 3]
```diff
@@ -139,13 +137,19 @@ func (d *sentinelProviderFilter) Invoke(ctx context.Context, invoker protocol.In
 	if b != nil {
 		// method blocked
 		return sentinelDubboProviderFallback(ctx, invoker, invocation, b)
+	} else {
+		defer methodEntry.Exit()
 	}
```
Comment (justxuewei) [inline filter/sentinel/filter.go:142]: Ditto.

File: filter/sentinel/filter.go [hunk 4]
```diff
@@ -177,22 +187,27 @@ func (d *sentinelConsumerFilter) Invoke(ctx context.Context, invoker protocol.In
 	if b != nil {
 		// interface blocked
 		return sentinelDubboConsumerFallback(ctx, invoker, invocation, b)
+	} else {
+		defer interfaceEntry.Exit()
 	}
```
Comment (justxuewei) [inline filter/sentinel/filter.go:192]: Ditto.

File: filter/sentinel/filter.go [hunk 5]
```diff
@@ -177,22 +187,27 @@ func (d *sentinelConsumerFilter) Invoke(ctx context.Context, invoker protocol.In
...
-	ctx = context.WithValue(ctx, InterfaceEntryKey, interfaceEntry)
 
 	methodEntry, b = sentinel.Entry(methodResourceName, sentinel.WithResourceType(base.ResTypeRPC),
 		sentinel.WithTrafficType(base.Outbound), sentinel.WithArgs(invocation.Arguments()...))
 	if b != nil {
 		// method blocked
 		return sentinelDubboConsumerFallback(ctx, invoker, invocation, b)
+	} else {
+		defer methodEntry.Exit()
 	}
```
Comment (justxuewei) [inline filter/sentinel/filter.go:201]: Ditto.

**Review by justxuewei — COMMENTED** [review]
Thanks! A few comments about code style.

In addition, I have another concern: `sentinel.TraceError` is called by `OnResponse()` previously, but now is called by `Invoke()`. **You changes the behavior of the filter.**

IMHO, you concern is that the context isn't updated with the latest data, right? So a possible fix could be:

```go
 func (fi *FilterInvoker) Invoke(ctx context.Context, invocation protocol.Invocation) protocol.Result { 
 	result := fi.filter.Invoke(ctx, fi.next, invocation)
 	// copy the context from the latest one
 	return fi.filter.OnResponse(ctx, result, fi.invoker, invocation) 
 }

**Comment by r27153733** [comment]
> Thanks! A few comments about code style.
> 
> In addition, I have another concern: `sentinel.TraceError` is called by `OnResponse()` previously, but now is called by `Invoke()`. **THE SIDE EFFECT: You changes the behavior of the filter.**
> 
> IMHO, your concern is that the context isn't updated with the latest data, right? So a possible fix could be:
> 
> ```go
>  func (fi *FilterInvoker) Invoke(ctx context.Context, invocation protocol.Invocation) protocol.Result { 
>  	result := fi.filter.Invoke(ctx, fi.next, invocation)
>  	// do something to copy the context from the latest one
>  	return fi.filter.OnResponse(ctx, result, fi.invoker, invocation) 
>  } 
> ```
> 
> Wdyt?
> 
> UPDATE 1: Maybe the above way is not easy to implement. But the main idea is to share the context between `fi.filter.Invoke()` and `fi.filter.OnResponse()`.

我也想在fi.filter.Invoke() 和 fi.filter.OnResponse()中共享上下文，很遗憾的是修改FilterInvoker是不够的，必须修改Filter接口的定义。
我曾经想从java中找到灵感，但是...

**Review by justxuewei — APPROVED** [review]
This looks good to me overall. For what I mentioned eariler, I found [a post](https://medium.com/@bserdar_67914/gos-context-how-can-i-copy-the-values-stored-in-a-context-cf17036ba1ea) introducing a wrapped context, which could be worth considering.

**Pull request merged**